In [105]:
import numpy as np
import time

Versão Vetorizada


$MSE(\hat{w})=\frac{1}{N}(y-\hat{\mathbf{w}}^T\mathbf{x})^T(y-\hat{\mathbf{w}}^T\mathbf{x})$

In [8]:
def compute_mse_vectorized(w,X,Y):
    #Y - Y_chapeu (predicao)
    res = Y - np.dot(X,w)
    totalError = np.dot(res.T,res)    
    return totalError / float(len(Y))#media da soma dos erros ao quadrado

In [108]:
def step_gradient_vectorized(w_current,X,Y,learningRate):
    #Y - Y_chapeu (predicao)
    res = Y - np.dot(X,w_current)
    
    #novo gradiente
    new_grad = np.multiply(-2/len(X),np.dot(X.T, res)) #usando a mesma ideia do vídeo do lab passado, multiplicando por -2/N
     
    #novos valores dos parametros
    new_w = np.zeros((len(new_grad),1))
    for i in range(len(new_grad)):
        new_w[i,0] = w_current[i] - (learningRate * new_grad[i])
        
    return [new_w,new_grad]

In [113]:
def gradient_descent_runner_vectorized(starting_w, X,Y, learning_rate, epsilon):
    w = starting_w
    grad = np.array([np.inf,np.inf, np.inf,np.inf,np.inf,np.inf])
    i = 0
    while (np.linalg.norm(grad)>=epsilon):
        w, grad = step_gradient_vectorized(w, X, Y, learning_rate)
        if i % 1000 == 0:
            print("MSE na iteração {0} é de {1}".format(i,compute_mse_vectorized(w, X, Y)))
        i+= 1
    return w

In [117]:
points = np.genfromtxt("sample_treino.csv", delimiter=",", skip_header=1)
points = np.c_[np.ones(len(points)),points]
X = points[:,:6]
Y = points[:,6][:,np.newaxis]

init_w = np.zeros((6,1))

learning_rate = 0.0001

epsilon = 0.285
print("Starting gradient descent at w0 = {0}, w1 = {1}, w2 = {2}, w3 = {3}, w4 = {4}, w5 = {5}, error = {6}".format(init_w[0], init_w[1],init_w[2], init_w[3],init_w[4], init_w[5], compute_mse_vectorized(init_w,X,Y)))
print("Running...")
tic = time.time()
w = gradient_descent_runner_vectorized(init_w, X,Y, learning_rate, epsilon)
toc = time.time()
print("Gradiente descendente convergiu com w0 = {0}, w1 = {1}, w2 = {2}, w3 = {3}, w4 = {4}, w5 = {5}, error = {6}".format(w[0], w[1],w[2], w[3],w[4], w[5], compute_mse_vectorized(w,X,Y)))
print("Versão vetorizada rodou em: " + str(1000*(toc-tic)) + " ms")

Starting gradient descent at w0 = [0.], w1 = [0.], w2 = [0.], w3 = [0.], w4 = [0.], w5 = [0.], error = [[54.47995386]]
Running...
MSE na iteração 0 é de [[48.42204544]]
MSE na iteração 1000 é de [[0.47897163]]
MSE na iteração 2000 é de [[0.46681063]]
Gradiente descendente convergiu com w0 = [0.03378612], w1 = [0.16605965], w2 = [0.20216232], w3 = [0.19079395], w4 = [0.28206898], w5 = [0.10360042], error = [[0.46059511]]
Versão vetorizada rodou em: 108.09540748596191 ms


Calculando os coeficientes usando o Scikit Learn (código disponível no notebook RegressaoMultiplaScikit.ipynb), obtém-se os seguintes resultados:
w0 = 0.58891358 , w1 = 0.1723595  , 0.13712469 , w3 = 0.15369677 , w4 = 0.33858381 , w5 = 0.05214724, error = 0.46

Para poder aproximar o erro desta solução à do Scikit, usei o epsilon = 0.285 (após alguns testes para definir o mais próximo).
Pode-se ver que os valores não são iguais, mas se aproximam bastante, tendo apenas o w0 como mais destoante, porém o obtido foi menor.